# Exploratory Data Analysis (EDA)

> Insurance Data Warehouse -- Bowtie Life Insurance Technical Assessment

**Objective**: Understand the data before building dbt models and SQL queries.

**Data Sources** (loaded into `raw` schema on PostgreSQL):
- `raw_policy` -- Insurance policies (6,583 rows)
- `raw_invoice` -- Billing invoices (9,646 rows)
- `raw_claim` -- Reimbursement claims (791 rows)

**Linking key**: `policy_number` connects all 3 tables.

In [2]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine("postgresql://postgres:postgres@localhost:5432/insurance_dwh")

def query(sql):
    """Helper to run SQL and return a DataFrame."""
    return pd.read_sql(sql, engine)

print("Connected to insurance_dwh")

Connected to insurance_dwh


---
## 1. Row Counts & Schema Overview

In [8]:
for table in ["raw_policy", "raw_invoice", "raw_claim"]:
    df = query(f"SELECT * FROM raw.{table} LIMIT 5")
    count = query(f"SELECT COUNT(*) AS cnt FROM raw.{table}")
    print(f"\n=== raw.{table} ({count['cnt'][0]} rows) ===")
    print(f"Columns: {list(df.columns)}")
    display(df)


=== raw.raw_policy (6583 rows) ===
Columns: ['id', 'policy_number', 'user_id', 'application_id', 'product', 'issue_date', 'effective_date', 'insured_gender', 'insured_date_of_birth']


,id,policy_number,user_id,application_id,product,issue_date,effective_date,insured_gender,insured_date_of_birth
0,458,0001104173200,f8636edb-894d-4d32-ab42-e41514fafc24,896171e1-abc0-4893-921d-94e06f9e7c1d,vhis,2020-07-13T20:54:24.978941+08:00,2020-07-09T00:09:33.682007+08:00,female,1980-01-01
1,3,0000576481439,4018b7ec-e16d-48d8-9e53-c63b8e1fe327,1e41d261-343a-4a0c-90d0-7d12927ec39d,vhis,2019-03-27T18:18:32.551194+08:00,2019-03-27T18:16:06.082524+08:00,male,1988-03-26
2,4,0000483424269,69493d2a-eb80-44f6-b65d-d92d687af3d7,796c27da-237c-4b7e-9aeb-fc41e3ea064d,vhis,2019-03-28T19:01:19.271047+08:00,2019-03-23T18:54:12.06124+08:00,female,2000-03-22
3,5,0001905133426,04bb1b46-f6a9-4af9-8f32-ccd59c847c19,4421a5a3-c0fa-41fe-9385-33360263ac2d,vhis,2019-03-29T13:47:11.49689+08:00,2019-03-29T13:45:39.927479+08:00,male,1998-03-09
4,6,0000971249312,04bb1b46-f6a9-4af9-8f32-ccd59c847c19,56cfe4cd-163a-4f78-81d9-e1caf00f107f,vhis,2019-03-29T15:49:37.38509+08:00,2019-03-29T15:40:52.883809+08:00,male,1993-03-28



=== raw.raw_invoice (9646 rows) ===
Columns: ['id', 'invoice_type', 'policy_number', 'coverage_start_date', 'coverage_end_date', 'due_date', 'status', 'pre_levy_amount', 'total_amount', 'refund_date', 'charge_date']


,id,invoice_type,policy_number,coverage_start_date,coverage_end_date,due_date,status,pre_levy_amount,total_amount,refund_date,charge_date
0,d1bca035-9f07-4d76-a6b1-9b9e2c444e77,regular,0002141605481,2019-09-06T11:32:12.803058+08:00,2019-10-06T11:32:12.803058+08:00,2019-09-06T11:32:12.803058+08:00,free,0.00,0.00,None,NaN
1,557fe953-50ab-4277-bb53-0978149a3e0e,regular,0001092357722,2021-05-16T14:31:12.603067+08:00,2021-06-16T14:31:12.603067+08:00,2021-05-16T14:31:12.603067+08:00,paid,200.00,200.20,None,2021-05-17T12:00:17.092921+08:00
2,dbafbe6e-55f9-4fd4-b9a6-3defee5f3b01,regular,0001213746921,2019-09-05T17:34:38.630843+08:00,2019-10-05T17:34:38.630843+08:00,2019-09-05T17:34:38.630843+08:00,paid,1034.00,1034.62,None,2019-09-06T12:00:46.03508+08:00
3,058d8298-b96f-432e-b510-77661b4ec762,regular,0000799162201,2019-09-06T11:08:07.401641+08:00,2019-10-06T11:08:07.401641+08:00,2019-09-06T11:08:07.401641+08:00,paid,104.00,104.06,None,2019-09-06T12:00:49.071365+08:00
4,6d96f276-caa0-44e8-8353-4ebc879295d4,regular,0001661166278,2019-10-20T10:26:41.209538+08:00,2019-11-20T10:26:41.209538+08:00,2019-10-20T10:26:41.209538+08:00,paid,191.00,191.11,None,2019-10-20T12:00:28.545709+08:00



=== raw.raw_claim (791 rows) ===
Columns: ['id', 'type', 'status', 'policy_number', 'submit_date', 'payment_date', 'admission_date', 'total_billed_amount', 'total_base_payable_amount']


,id,type,status,policy_number,submit_date,payment_date,admission_date,total_billed_amount,total_base_payable_amount
0,00060102-fb3c-4e77-bad2-2e20ccbdf012,reimbursement,open,0001617436567,2019-12-23T15:35:52.417933+08:00,None,NaN,0,0
1,0008f3b3-057b-43eb-827a-7efbd93b8045,reimbursement,open,0002115730101,2021-11-22T14:47:11.840617+08:00,None,NaN,0,0
2,00a17afd-fd90-4f10-929f-90c49b6a1a2d,reimbursement,withdrawn,0001235638810,2021-10-11T10:53:47.171607+08:00,None,NaN,0,0
3,00b49aab-7f67-4d65-9cab-88eb125290d7,reimbursement,open,0000963774004,2022-02-10T20:04:55.405226+08:00,None,NaN,0,0
4,014c9653-1c1b-4e69-9253-80fb90fa0d8f,reimbursement,payment-requested,0001315900696,2022-03-06T17:45:55.910154+08:00,None,2022-03-06,0,0


---
## 2. raw_policy -- Deep Dive

In [9]:
# Unique policy numbers and unique users
query("""
    SELECT
        COUNT(*)                          AS total_rows,
        COUNT(DISTINCT policy_number)     AS unique_policies,
        COUNT(DISTINCT user_id)           AS unique_users
    FROM raw.raw_policy
""")

,total_rows,unique_policies,unique_users
0,6583,6583,166


In [10]:
# How many policies does each user hold?
query("""
    SELECT
        policies_per_user,
        COUNT(*) AS user_count
    FROM (
        SELECT user_id, COUNT(*) AS policies_per_user
        FROM raw.raw_policy
        GROUP BY user_id
    ) sub
    GROUP BY policies_per_user
    ORDER BY policies_per_user
""")

,policies_per_user,user_count
0,1,76
1,2,27
2,3,14
3,4,6
4,5,5
5,6,3
6,7,3
7,8,3
8,9,4
9,10,1


In [13]:
# Product name distribution
query("""
    SELECT product, COUNT(*) AS cnt
    FROM raw.raw_policy
    GROUP BY product
    ORDER BY cnt DESC
""")

,product,cnt
0,outpatient,6050
1,vhis,229
2,vision-care,63
3,life,54
4,medical-rider,45
5,critical-illness,41
6,preventive-care,37
7,medical,33
8,accidental-medical,31


In [3]:
# Date ranges: when were policies effective?
query("""
    SELECT
        MIN(effective_date::date) AS earliest_effective,
        MAX(effective_date::date) AS latest_effective
    FROM raw.raw_policy
""")

,earliest_effective,latest_effective
0,2019-01-01,2022-12-16


In [16]:
# Null check across all columns
df_policy = query("SELECT * FROM raw.raw_policy")
df_policy.isnull().sum().to_frame("null_count")

,null_count
id,0
policy_number,0
user_id,0
application_id,0
product,0
issue_date,0
effective_date,0
insured_gender,2
insured_date_of_birth,3


---
## 3. raw_invoice -- Deep Dive

In [18]:
# Unique invoice numbers and linked policies
query("""
    SELECT
        COUNT(*)                          AS total_rows,
        COUNT(DISTINCT policy_number)     AS policies_with_invoices
    FROM raw.raw_invoice
""")

,total_rows,policies_with_invoices
0,9646,855


In [20]:
# Invoice status distribution
query("""
    SELECT status, COUNT(*) AS cnt
    FROM raw.raw_invoice
    GROUP BY status
    ORDER BY cnt DESC
""")

,status,cnt
0,paid,8176
1,open,1026
2,void,170
3,free,156
4,refunded,118


In [21]:
# Premium stats: pre_levy_amount (net premium) vs total_amount (includes IA levy)
query("""
    SELECT
        MIN(pre_levy_amount::numeric)  AS min_pre_levy,
        AVG(pre_levy_amount::numeric)  AS avg_pre_levy,
        MAX(pre_levy_amount::numeric)  AS max_pre_levy,
        MIN(total_amount::numeric)     AS min_total,
        AVG(total_amount::numeric)     AS avg_total,
        MAX(total_amount::numeric)     AS max_total
    FROM raw.raw_invoice
""")

,min_pre_levy,avg_pre_levy,max_pre_levy,min_total,avg_total,max_total
0,0.0,178.470708,7383.0,0.0,178.590537,7390.38


In [30]:
# Invoice date range
query("""
    SELECT
        MIN(charge_date::date) AS earliest_invoice,
        MAX(charge_date::date) AS latest_invoice,
        MIN(due_date::date)     AS earliest_due,
        MAX(due_date::date)     AS latest_due,
        MIN(refund_date::date) AS earliest_refund,
        MAX(refund_date::date) AS latest_refund
    FROM raw.raw_invoice
""")

,earliest_invoice,latest_invoice,earliest_due,latest_due,earliest_refund,latest_refund
0,2019-03-27,2022-12-21,2019-03-17,2022-12-21,2019-03-28,2022-11-16


In [24]:
# Null check
df_invoice = query("SELECT * FROM raw.raw_invoice")
df_invoice.isnull().sum().to_frame("null_count")

,null_count
id,0
invoice_type,0
policy_number,0
coverage_start_date,2
coverage_end_date,2
due_date,0
status,0
pre_levy_amount,0
total_amount,0
refund_date,9528


---
## 4. raw_claim -- Deep Dive

In [26]:
# Unique claims and linked policies
query("""
    SELECT
        COUNT(*)                          AS total_rows,
        COUNT(DISTINCT policy_number)     AS policies_with_claims
    FROM raw.raw_claim
""")

,total_rows,policies_with_claims
0,791,124


In [27]:
# Claim status distribution
query("""
    SELECT status, COUNT(*) AS cnt
    FROM raw.raw_claim
    GROUP BY status
    ORDER BY cnt DESC
""")

,status,cnt
0,open,446
1,payment-requested,117
2,accepted,83
3,paid,55
4,withdrawn,47
5,updated,23
6,declined,6
7,escalated,5
8,canceled,3
9,update-requested,3


In [29]:
# Claim amounts & reimbursement amounts
# total billed amount is the sticker price billed by doctor
# total base payable amount is the actual amount that claimed by the insured.
query("""
    SELECT
        ROUND(MIN(total_billed_amount::numeric), 2)         AS min_billed,
        ROUND(AVG(total_billed_amount::numeric), 2)         AS avg_billed,
        ROUND(MAX(total_billed_amount::numeric), 2)         AS max_billed,
        ROUND(MIN(total_base_payable_amount::numeric), 2)   AS min_payable,
        ROUND(AVG(total_base_payable_amount::numeric), 2)   AS avg_payable,
        ROUND(MAX(total_base_payable_amount::numeric), 2)   AS max_payable
    FROM raw.raw_claim
""")

,min_billed,avg_billed,max_billed,min_payable,avg_payable,max_payable
0,0.0,6189.67,1331000.0,0.0,6643.1,2000000.0


In [31]:
# Claim date ranges
query("""
    SELECT
        MIN(submit_date::date)    AS earliest_submission,
        MAX(submit_date::date)    AS latest_submission,
        MIN(payment_date::date) AS earliest_payment,
        MAX(payment_date::date) AS latest_payment
    FROM raw.raw_claim
""")

,earliest_submission,latest_submission,earliest_payment,latest_payment
0,2019-03-29,2022-11-29,2019-09-25,2022-11-24


In [32]:
# Null check
df_claim = query("SELECT * FROM raw.raw_claim")
df_claim.isnull().sum().to_frame("null_count")

,null_count
id,0
type,0
status,0
policy_number,0
submit_date,0
payment_date,735
admission_date,446
total_billed_amount,0
total_base_payable_amount,0


---
## 5. Cross-Table Relationships

Key questions:
- Do all invoice `policy_number`s exist in the policy table?
- Do all claim `policy_number`s exist in the policy table?
- What % of policies have invoices? What % have claims?

In [33]:
# Orphan check: invoices or claims referencing non-existent policies
query("""
    SELECT 'invoice' AS source,
           COUNT(*) AS orphan_count
    FROM raw.raw_invoice i
    LEFT JOIN raw.raw_policy p ON i.policy_number = p.policy_number
    WHERE p.policy_number IS NULL

    UNION ALL

    SELECT 'claim' AS source,
           COUNT(*) AS orphan_count
    FROM raw.raw_claim c
    LEFT JOIN raw.raw_policy p ON c.policy_number = p.policy_number
    WHERE p.policy_number IS NULL
""")

,source,orphan_count
0,invoice,2
1,claim,0


In [34]:
# Coverage: what % of policies have at least one invoice / claim?
query("""
    SELECT
        (SELECT COUNT(DISTINCT policy_number) FROM raw.raw_policy)   AS total_policies,
        (SELECT COUNT(DISTINCT policy_number) FROM raw.raw_invoice)  AS policies_with_invoices,
        (SELECT COUNT(DISTINCT policy_number) FROM raw.raw_claim)    AS policies_with_claims
""")

,total_policies,policies_with_invoices,policies_with_claims
0,6583,855,124


---
## 6. New vs. Returning Policies (Q1 preview)

> Q1 asks: "average net premium comparing policies that are new vs. returning."
> 
> We need to figure out how to define **new** vs **returning**.
> A user's **first policy** = new; any subsequent policy = returning.

In [35]:
# How many users have more than one policy? (potential "returning" customers)
query("""
    SELECT
        CASE
            WHEN policy_count = 1 THEN 'single_policy'
            ELSE 'multi_policy'
        END AS user_type,
        COUNT(*) AS user_count
    FROM (
        SELECT user_id, COUNT(*) AS policy_count
        FROM raw.raw_policy
        GROUP BY user_id
    ) sub
    GROUP BY user_type
""")

,user_type,user_count
0,multi_policy,90
1,single_policy,76


---
## 7. Key Findings & Special Notes

### 7.1 raw_policy (6,583 rows, 9 columns)

| Finding | Value |
|---|---|
| Unique policies | 6,583 (`policy_number` is unique -- no duplicates) |
| Unique users | **166** (~40 policies per user on average) |
| Top product | `outpatient` dominates with **6,050 (92%)** |
| Other products | vhis (229), vision-care (63), life (54), medical-rider (45), critical-illness (41), preventive-care (37), medical (33), accidental-medical (31) |
| Date range | 2019-01-01 to 2022-12-16 (~4 years) |
| Nulls | `insured_gender` (2), `insured_date_of_birth` (3) -- negligible |

**Outlier alert**: One user holds **5,380 policies** (82% of all policies). The next highest is 155. This is almost certainly a test/system account and should be flagged or excluded in dbt staging models.

---

### 7.2 raw_invoice (9,646 rows, 11 columns)

| Finding | Value |
|---|---|
| Policies with invoices | **855** out of 6,583 (only **13%**) |
| Status breakdown | paid (8,176), open (1,026), void (170), free (156), refunded (118) |
| Avg pre-levy premium | **$178.47** |
| Max premium | $7,383.00 |
| Date range (charge_date) | 2019-03-27 to 2022-12-21 |
| Key nulls | `refund_date` (9,528 -- expected), `charge_date` (1,351), `coverage_start/end_date` (2 each) |

The low 13% invoice coverage likely reflects the large number of free/trial outpatient policies from the outlier user.

**Column mapping for Q1**:
- `pre_levy_amount` = net premium (before IA levy) -- **this is what Q1 asks about**
- `total_amount` = gross premium (pre_levy_amount + 0.1% Insurance Authority levy)

---

### 7.3 raw_claim (791 rows, 9 columns)

| Finding | Value |
|---|---|
| Policies with claims | **124** out of 6,583 (1.9%) |
| Top statuses | open (446), payment-requested (117), accepted (83), paid (55), withdrawn (47) |
| Avg billed amount | **$6,189.67** |
| Max payable amount | **$2,000,000** |
| Key nulls | `payment_date` (735 -- most claims not yet paid), `admission_date` (446) |

Note: `total_base_payable_amount` can exceed `total_billed_amount` (max $2M vs $1.33M) -- worth investigating but not necessarily an error.

**Column mapping**:
- `total_billed_amount` = the sticker price billed by the healthcare provider
- `total_base_payable_amount` = the amount payable under the insurance policy

---

### 7.4 Cross-Table Relationships

| Check | Result |
|---|---|
| Orphan invoices (no matching policy) | **2** |
| Orphan claims (no matching policy) | **0** |
| Total policies | 6,583 |
| Policies with at least 1 invoice | 855 (13%) |
| Policies with at least 1 claim | 124 (1.9%) |

---

### 7.5 New vs. Returning Users (Q1 Preview)

| User type | Count |
|---|---|
| Single-policy users (new only) | **76** |
| Multi-policy users (have returning policies) | **90** |

**Definition for Q1**: Rank each user's policies by `effective_date` using `ROW_NUMBER()`. Rank 1 = **new**, rank 2+ = **returning**.

---

### 7.6 Key Decisions for dbt Models

1. **Outlier user** (5,380 policies) -- flag or exclude in staging
2. **Type casting** -- all dates stored as text with timezone offsets, amounts as text; cast in staging models
3. **Null handling** -- `payment_date` (735 nulls), `admission_date` (446 nulls), `refund_date` (9,528 nulls) are expected nulls, not data quality issues
4. **Invoice filter for Q1** -- consider using only `status = 'paid'` invoices when computing "premium received"
5. **Orphan invoices** -- filter the 2 orphan records in staging with a referential integrity test
6. **New vs returning** -- use `ROW_NUMBER() OVER (PARTITION BY user_id ORDER BY effective_date)` to classify